### ISSUE TITLE
New rule: all sealed class children must be data classes.

### ISSUE THREAD

**Issue by VovaStelmashchuk** [detekt/detekt#6630]
Add a new rule all children of `sealed class` must be data classes.  

The sealed class is usually used to describe the system or feature states. 

Also, the rule prevents the creation of an open class as a child for the `sealed class`, it reduces the count of the potential of unnecessary abstraction in the system.

Also, the rule fixes the potential bug with comparing states because the data classes have auto-generated equals and hashcode methods.

## Expected Behavior of the rule
Correct code:
```kotlin
sealed class Foo {
    object Bar : Foo()
    data class Baz(val i: Int) : Foo()
}

```

Wrong code:
```kotlin
sealed class Foo {
    object Bar : Foo()
    open class Baz : Foo()

    open class Baz2(val test: Int) : Baz()
}
```

**Comment by BraisGabin** [comment]
To be more specific (because its the only way to implement the rule). A `sealed class` can only be extended by a:

- `data class`
- `data object` (Better `toString` than a regular `object`)
- `sealed class` (The "tree" doesn't need to have one single layer)

And, what about the `sealed inerfaces`? Right now I'm not in favor nor against of this rule but we need to make it more specific to decide what to do with it.

**Comment by VovaStelmashchuk** [comment]
Yep, you catch the point.
`value class` is not the case. A value class cannot extend other classes. 

About `sealed interface` not sure we need to handle the case because it looks pretty standard for me to have an open class as a child of a sealed interface. 

Having a normal class as the children for the sealed class does not look good for me because the developers usually use the children of the sealed class to describe the states. The state classes are often a part of some checkers, like an equal or contains, etc.

```kotlin
val queue = PriorityQueue<SealedClass>()

if (queue.contains(foo)) {
//Do something 
}
```

In case foo is an instance of a normal class without implementing equals and hashcode the contains login doesn't work as expected.

**Comment by cortinico** [comment]
> it reduces the count of the potential of unnecessary abstraction in the system.

I'm not super convinced by this rule. What if I'm in a constrained environment and I don't want to use a `data class` but just a `class` to don't have to deal with all the extra methods generated for data classes?

**Comment by VovaStelmashchuk** [comment]
I didn't think to include this rule by default. My idea is to have this rule excluded by default. 

The developers be able to enable it for all classes that are responsible for UiState or just for all sealed classes or whatever. 

If it's not consistent with the detekt rules, it's totally ok for me. I can wrap the rule as a detekt plugin and maintain the project myself.

**Comment by BraisGabin** [comment]
Closing for the reasons described on the PR.
